packages/labs/src/components/omnibox/omnibox.tsx

```diff
@@ -0,0 +1,197 @@
+     * will not actually close itself until that prop becomes `false`.
+     */
+    onClose?: (event?: React.SyntheticEvent<HTMLElement>) => void;
+
+    /** Props to spread to `Overlay`. Note that `content` cannot be changed. */
+    overlayProps?: Partial<IOverlayProps>;
+
+    /**
+     * Whether the filtering state should be reset to initial when an item is selected
+     * (immediately before `onItemSelect` is invoked). The query will become the empty string
+     * and the first item will be made active.
+     * @default false
+     */
+    resetOnSelect?: boolean;
+}
+
+export interface IOmniboxState<T> extends IOverlayableProps, IBackdropProps {
+    activeItem?: T;
+    query?: string;
+}
+
+@PureRender
+export class Omnibox<T> extends AbstractComponent<IOmniboxProps<T>, IOmniboxState<T>> {
+    public static displayName = "Blueprint.Omnibox";
+
+    public static ofType<T>() {
+        return Omnibox as new () => Omnibox<T>;
+    }
+
+    public state: IOmniboxState<T> = {
+        query: "",
+    };
+
+    private TypedQueryList = QueryList.ofType<T>();
+    private queryList: QueryList<T>;
+    private refHandlers = {
+        queryList: (ref: QueryList<T>) => this.queryList = ref,
+    };
+
+    public render() {
+        // omit props specific to this component, spread the rest.
+        const { isOpen, itemRenderer, inputProps, noResults, overlayProps, ...props } = this.props;
+        return <this.TypedQueryList
+            {...props}
+            activeItem={this.state.activeItem}
+            onActiveItemChange={this.handleActiveItemChange}
+            onItemSelect={this.handleItemSelect}
+            query={this.state.query}
+            ref={this.refHandlers.queryList}
+            renderer={this.renderQueryList}
+        />;
+    }
+
+    public componentWillReceiveProps(nextProps: IOmniboxProps<T>) {
+        const { isOpen } = nextProps;
+        const canClearQuery = !this.props.isOpen && isOpen && this.props.resetOnSelect;
+
+        this.setState({
+            activeItem: canClearQuery ? this.props.items[0] : this.state.activeItem,
+            query: canClearQuery ? "" : this.state.query,
+        });
+    }
+
+    private renderQueryList = (listProps: IQueryListRendererProps<T>) => {
+        const { inputProps = {}, isOpen, overlayProps = {} } = this.props;
+        const { query } = this.state;
+        const { ref, ...htmlInputProps } = inputProps;
+        const { handleKeyDown, handleKeyUp } = listProps;
+
+        return (
+            <Overlay
+                hasBackdrop={true}
+                {...overlayProps}
+                isOpen={isOpen}
+                className={classNames(overlayProps.className, `${Classes.OMNIBOX}-overlay`)}
+                onClose={this.handleOverlayClose}
+            >
+                <div
+                    className={classNames(listProps.className, Classes.OMNIBOX)}
+                    onKeyDown={isOpen && query.length > 0 && handleKeyDown}
```

Comment:
> Subsidiary question, is this syntax ever okay?

It may work, but I don't find it particularly clear. I like everything in a boolean expression to be an actual boolean check.